core/src/main/java/org/apache/calcite/rel/rules/JoinToCorrelateRule.java

```diff
@@ -77,10 +83,19 @@ protected JoinToCorrelateRule(RelFactories.FilterFactory filterFactory) {
     this(RelBuilder.proto(Contexts.of(filterFactory)));
   }
 
+  /**
+   * Creates a JoinToCorrelateRule for a certain sub-class of
+   * {@link org.apache.calcite.rel.core.Join}
+   */
+  public JoinToCorrelateRule(Class<? extends Join> clazz,
```

Comment:
If we can treat any kind of join with this rule then we shouldn't have to perform casts afterwards. I am OK with adding such a constructor if we can achieve that.

Reply:
I think we should be a bit conservative regarding the public API. The goal of this PR is to allow semi joins to be transformed to correlates. This can be also achieved by keeping this constructor private so I would propose to reduce its visibility.